ext/odbc/php_odbc.c

```diff
@@ -49,23 +51,175 @@
+	zend_throw_error(NULL, "Cannot directly construct ODBC\\Connection, use odbc_connect() or odbc_pconnect() instead");
+	return NULL;
+}
+
+static void odbc_link_free(odbc_link *link)
+{
+	ZEND_ASSERT(link->connection && "link has already been closed");
+
+	if (link->persistent) {
+		zend_hash_del(&EG(persistent_list), link->hash);
+	} else {
+		close_results_with_connection(link->connection);
+
+		zend_hash_del(&ODBCG(connections), link->hash);
+
+		/* If aborted via timer expiration, don't try to call any unixODBC function */
+		if (!(PG(connection_status) & PHP_CONNECTION_TIMEOUT)) {
+			safe_odbc_disconnect(link->connection->hdbc);
+			SQLFreeConnect(link->connection->hdbc);
+			SQLFreeEnv(link->connection->henv);
+		}
+		efree(link->connection);
+		ODBCG(num_links)--;
+	}
+
+	link->connection = NULL;
+
+	zend_string_release(link->hash);
+	link->hash = NULL;
+}
+
+static void odbc_connection_free_obj(zend_object *obj)
+{
+	odbc_link *link = odbc_link_from_obj(obj);
+
+	if (!link->connection) {
+		zend_object_std_dtor(&link->std);
+		return;
+	}
+
+	odbc_link_free(link);
+	zend_object_std_dtor(&link->std);
+}
+
+static inline odbc_result *odbc_result_from_obj(zend_object *obj) {
+	return (odbc_result *)((char *)(obj) - XtOffsetOf(odbc_result, std));
+}
+
+#define Z_ODBC_RESULT_P(zv) odbc_result_from_obj(Z_OBJ_P(zv))
+
+static zend_object *odbc_result_create_object(zend_class_entry *class_type) {
+	odbc_result *intern = zend_object_alloc(sizeof(odbc_result), class_type);
+
+	zend_object_std_init(&intern->std, class_type);
+	object_properties_init(&intern->std, class_type);
+	intern->std.handlers = &odbc_result_object_handlers;
+
+	return &intern->std;
+}
+
+static zend_function *odbc_result_get_constructor(zend_object *object) {
+	zend_throw_error(NULL, "Cannot directly construct ODBC\\Result, use an appropriate odbc_* function instead");
+	return NULL;
+}
+
+static void odbc_result_free(odbc_result *res)
+{
+	ZEND_ASSERT(res->conn_ptr && "result has already been closed");
+
+	if (res->values) {
+		for (int i = 0; i < res->numcols; i++) {
+			if (res->values[i].value) {
+				efree(res->values[i].value);
+			}
+		}
+		efree(res->values);
+		res->values = NULL;
+		res->numcols = 0;
+	}
+	/* If aborted via timer expiration, don't try to call any unixODBC function */
+	if (res->stmt && !(PG(connection_status) & PHP_CONNECTION_TIMEOUT)) {
+#if defined(HAVE_SOLID) || defined(HAVE_SOLID_30) || defined(HAVE_SOLID_35)
+		SQLTransact(res->conn_ptr->henv, res->conn_ptr->hdbc,
+					(SQLUSMALLINT) SQL_COMMIT);
+#endif
+		SQLFreeStmt(res->stmt,SQL_DROP);
+		/* We don't want the connection to be closed after the last statement has been closed
+		 * Connections will be closed on shutdown
+		 */
+		res->stmt = NULL;
+	}
+	if (res->param_info) {
+		efree(res->param_info);
+		res->param_info = NULL;
+	}
+	res->conn_ptr = NULL;
+}
 
-static int le_result, le_conn, le_pconn;
+static void odbc_result_free_obj(zend_object *obj) {
```

Comment:
Uh I should have noticed this :/ You are completely right!

Reply:
I also wonder if it makes sense to store the results per connection instead of using a module global.
I haven't tried this, but I think that might be cleaner and would allow you to at least simplify `close_results_with_connection` to no longer use a O(n) loop.

Reply:
Oh yes, that sounds a lot better than the current mess. I'll give it a try